### ISSUE TITLE
Context policies are not correctly resolved for nested dependencies (version 5.8.1)

### ISSUE THREAD

**Issue by Evor78** [unitycontainer/unity#210]
When using the reflection (stage = pre-build) and wireup (stage = initialization) build strategies, the latter can't resolve policies stored in the policy list during the pre-build stage. This happens only if the instance is being resolved from another instance.

I've attached the sample console project reproducing the problem using the EventBroker extension.
The sample works correctly in Unity Container v. 2.1.505.2 (also included).

Notes about the sample project:
- The application creates 1 event publisher and 2 subscriber instances;
- The subscriber instance which is resolved directly form the container successfully wires up its event handler, but another instance which is resolved from the wrapper is unable to wire up its handler, because the PolicyList does not return correct (and really existing) policy;
- See BrokerReflectionStrategy, line 26 (add policy);
- See BrokerWireUpStrategy, line 20 (policy retrieval);

Thanks.

[EventBrokerTest.zip](https://github.com/unitycontainer/unity/files/1860133/EventBrokerTest.zip)
![with unitycontainer v5x](https://user-images.githubusercontent.com/37900953/38089992-e2a2bdc2-3369-11e8-901e-4d2b6936240a.png)
![with unitycontainer v2 1 505 2](https://user-images.githubusercontent.com/37900953/38089997-e7a29cd4-3369-11e8-8cb7-facd196ee264.png)

**Comment by ENikS** [comment]
@Evor78  Container is in 5.8.* now.
You are targeting code from 5/11/2011. A lot has changed in last 7 years. 
You would have to upgrade to latest behavior if you want to keep up...

**Comment by Evor78** [comment]
No!
I'm trying to use the 5.8.1.0 version and it fails!
Could you please at least compile and run the sample?

Thanks.

**Comment by ENikS** [comment]
Unfortunately I can't download and test your sample. Could you provide a unit test like this:
```C#
        [TestMethod]
        public void SelectConstructorWithNoParameters()
        {
            // Setup
            ...

            // Act
           ...

            // Verify
            Assert....
        }
```

**Comment by Evor78** [comment]
Hmm...
I guess I need to reimplement it without EventBroker, because the latter is rather big.
I'll try to do it then.

**Comment by Evor78** [comment]
Here is the test class (see comments inline):

[TestClass]
    public class UcPolicyTest
    {
        /// <summary>
        /// Tests policies with the class that is resolved directly from the container.
        /// </summary>
        [TestMethod]
        public void PolicyTest()
        {
            var container = new UnityContainer();
            container.AddNewExtension<MyExtension>();

            var instance = container.Resolve<MyClass>();
            instance.DoAssert();
        }

        /// <summary>
        /// Tests policies with the class that is resolved from the container 
        /// indirectly using the wrapper class.
        /// This test passes in Unity Container 2.1.505.2, but fails in 5.8.1
        /// </summary>
        [TestMethod]
        public void PolicyWrapperTest()
        {
            var container = new UnityContainer();
            container.AddNewExtension<MyExtension>();

            var wrapperInstance = container.Resolve<MyClassWrapper>();
            wrapperInstance.DoAssert();
        }

        #region Classes

        public class MyClassWrapper
        {
            private readonly MyClass wrappedInstance;

            public MyClassWrapper(MyClass wrappedInstance)
            {
                this.wrappedInstance = wrappedInstance;
            }

            public void Do()
            {
                wrappedInstance.Do();
            }

            public void DoAssert()
            {
                wrappedInstance.DoAssert();
            }
        }

        public class MyClass
        {
            private bool wasCalled;

            public void Do()
            {
                wasCalled = true;
            }

            public void DoAssert()
            {
                Assert.IsTrue(wasCalled, "The Do() method was not called from the wireup strategy!");
            }
        }

        public class MyExtension : UnityContainerExtension
        {
            protected override void Initialize()
            {
                Context.Strategies.Add(new PreCreationStrategy(), UnityBuildStage.PreCreation);
                Context.Strategies.Add(new WireupStrategy(), UnityBuildStage.Initialization);
            }
        }

        public class MyPolicy : IBuilderPolicy { }

        public class PreCreationStrategy : BuilderStrategy
        {
            public override void PreBuildUp(IBuilderContext context)
            {
                if (context.BuildKey.Type == typeof(MyClass))
                {
                    var policy = new MyPolicy();
                    context.Policies.Set(policy, context.BuildKey);
                }
                base.PreBuildUp(context);
            }
        }

        public class WireupStrategy : BuilderStrategy
        {
            public override void PreBuildUp(IBuilderContext context)
            {
                if (context.Existing is MyClass instance)
                {
                    var policy = context.Policies.Get<MyPolicy>(context.BuildKey);
                    if (policy != null)
                    {
                        instance.Do();
                    }
                }
                base.PreBuildUp(context);
            }
        }

        #endregion
    }

**Comment by ENikS** [comment]
Your issue is with [this change](https://github.com/unitycontainer/unity/wiki/Upgrading-to-Release-v5.2.1)Prior to 5.2.1 there were two registrations for each FromType and ToType. Now it is only FromType.

When you set your ``context.Policies.Set<IEventBrokerPolicy>(policy, context.OriginalBuildKey)`` on build type it only works for the type you register it for. 
```
c.RegisterType<ISubscriber, Subscriber>();
c.RegisterType<ISubscriberWrapper, SubscriberWrapper>();
```
These are completely independent now and have to common denomitantor. This is why it is not found by the extension.

**Comment by Evor78** [comment]
I don't think it's related to registration change.
The unit test I've provided does not contain any registrations (even no interfaces).

I think I've managed to find the possible root cause:
![putgetissue](https://user-images.githubusercontent.com/37900953/38147257-230d5958-345a-11e8-84b3-7358a45c3c36.png)

Could you please comment why I can't get back something that I put in a line earlier?
The unit test reproduces this issue.

Thanks for your patience.

**Comment by ENikS** [comment]
I would suggest you get local Build of Unity and add your unity test to one of the test projects. You will be able to check internals and see where your policies are placed and why it is not working.

**Comment by Evor78** [comment]
I've found a workaround by using the internal UnityContainer's SetPlicy/GetPolicy methods.

I'm not going to dig into your sources, guys, but there is a real bug. The UT proves it.
All I found is that calling context.Policies.Set() method spawns a new policy list that is isolated from the container. That doesn't work for chained dependencies. There are many implementations of the IPolicyList in your code, but only one of them does not use the container at all.
